src/lib.rs

```diff
@@ -511,37 +551,85 @@ impl Drop for CANSocket {
     }
 }
 
-/// CANFrame
+/// CANFrameStruct
 ///
 /// Uses the same memory layout as the underlying kernel struct for performance
 /// reasons.
 #[derive(Debug, Copy, Clone)]
 #[repr(C)]
-pub struct CANFrame {
+struct CANFrameStruct {
     /// 32 bit CAN_ID + EFF/RTR/ERR flags
     _id: u32,
 
-    /// data length. Bytes beyond are not valid
+    /// frame payload length in bytes (0 .. 8) for classical frame,
+    /// (0 .. 64) for FD frame
     _data_len: u8,
 
-    /// padding
-    _pad: u8,
+    /// padding for classical frame, additional
+    /// flags for CAN FD
+    _pad_flags: u8,
 
     /// reserved
     _res0: u8,
 
     /// reserved
     _res1: u8,
 
-    /// buffer for data
-    _data: [u8; 8],
+    /// buffer for data (classical or FD frame)
+    _data: [u8; CAN_FRAME_DATA_LEN_MAX]
```

Comment:
This is not an issue newly introduced with this PR, but does this accurately capture the behavior/memory layout induced by ```__attribute__((aligned(8)))``` in the original C header? If ```#[repr(C)]``` captures the default C behavior, the entire struct should have alignof = 4, which will probably be irrelevant for heap allocations (as the allocator just defaults to boundaries of 8 on most systems), but might be utilized by the compiler on the stack.. 

https://doc.rust-lang.org/reference/type-layout.html mentions that you could do ```#[repr(C, align(8))]``` on the struct, which, in combination with the padding, would probably always shift this field to the right offset?

Reply:
I don't really understand why the `__attribute__((aligned(8)))` in the original C header in the Linux source is at the `data` field and not at the entire `struct can_frame` (`canfd_frame`)?

What if the struct begins at an multiple of 4 address that's not multiple of 8? Would it just allocate 4 more bytes at the end of the `data` or would it make the `data` begins at the boundary of 8 (which would introduce a 4 bytes gap between `_res1` (`len8_dlc` respectively) and `data`?

Reply:
I would guess that the reason to align the data field is to be completely explicit about the intent; that you could expect the data to be aligned even if the rest of the struct changed in the future.

That said, clearly the existing Rust implementation is incorrect. That alignment must be made part of the Rust struct definition. I remember an old trick where you put a zero-length array into the struct of a type to align it, like:
```
pub struct CanFrame {
    _id: u32,
    _data_len: u8,
    _pad: u8,
    _res0: u8,
    _res1: u8,

    _alignment: [u64; 0], 
    _data: [u8; 8],
}
```
This appears to align the whole struct to 8-bytes, and not just the data field. But that's OK. It would work.

So then I wondered, "What would bindgen do?", and there it its:
```
#[repr(C)]
#[repr(align(8))]
#[derive(Debug, Copy, Clone)]
pub struct can_frame {
    pub can_id: canid_t,
    pub can_dlc: __u8,
    pub __pad: __u8,
    pub __res0: __u8,
    pub __res1: __u8,
    pub data: [__u8; 8usize],
}
```
It just aligns the whole struct, landing the `data` field at the right place. It's the same thing that the `_alignment` field does, but is more explicit, so preferable.

Reply:
@jreppnow Thanks for the link to the C standard. If you can ask your colleagues about the attribute placement it would be great.

I'd go for `#[repr(align(8))]` on the whole struct if there is no serious reason to do otherwise.